Add efficient zero tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64837

Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

anjali411 committed:
Add efficient zero tensors

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -166,6 +165,8 @@ enum class DispatchKey : uint8_t {
   // This is implemented at a dispatch level right before any backends run
   Negative,
 
+  ZeroTensor,
```
@bdhirsh dispatch keys out of space 🤣

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4574,6 +4579,8 @@
   device_check: NoCheck
   device_guard: False
 
+- func: efficientzerotensor(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
I hate this function name lol

anjali411 replied:
Alban and I were thinking of adding a kwarg argument to `torch.zeros` and expose it to users in that way. This is only a temp function to get the basic stuff working

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

albanD commented on BUILD.bazel:
```diff
@@ -136,6 +136,7 @@ genrule(
         "aten/src/ATen/RegisterQuantizedCPU.cpp",
         "aten/src/ATen/RegisterSparseCPU.cpp",
         "aten/src/ATen/RegisterSparseCsrCPU.cpp",
+        "aten/src/ATen/RegisterZeroTensor.cpp",
```
this will most likely have a fbcode only change

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -606,13 +606,40 @@ Tensor& floor_divide_(Tensor& self, const Tensor& other) {
 // TODO: Make this structured to undo the perf regression from native:: removal
 // in call here
 Tensor mul(const Tensor& self, const Scalar& other) {
+  // prob ok if self never requires grad
```
Not sure what this comment means? But no this should return a brand new Tensor.

anjali411 replied:
ahh yeah it's a stale comment. This is never called if `ZeroTensor` dispatch key is set.

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -606,13 +606,40 @@ Tensor& floor_divide_(Tensor& self, const Tensor& other) {
   return at::mul_out(self, wrapped_scalar_tensor(other), self); // redispatch!
 }
 
+void size_and_device_check(const Tensor& self, const Tensor& other) {
```
Is this ever used?

anjali411 replied:
yup it's called from `mul_zerotensor` and `add_zerotensor`

albanD replied:
Not sure how I missed that. ctrl+f failed me.

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

albanD commented on aten/src/ATen/ZeroTensorFallback.cpp:
```diff
@@ -0,0 +1,99 @@
+    m.impl("empty_strided", torch::CppFunction::makeFallthrough());
+    m.impl("mul.Scalar", torch::CppFunction::makeFallthrough());
+    // specific registeration in native_functions.yaml
+    // m.impl("mul.Tensor", torch::CppFunction::makeFallthrough());
```
Should this be removed?
If you want to document that some of these are specified in native functions, maybe just use a list? Or tell the reader to check that other file (so that the list doesn't go stale)?

I feel like having some as fallthrough and some as special definition in native_functions is a bit confusing. Should we just use one here?
If you prefer to keep both, it would be nice to have a comment here saying that we're doing both and the reader should not be surprised.

albanD commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -214,6 +214,10 @@ inline bool is_inference(const Tensor& tensor) {
 }
 
+inline bool is_zerotensor(const Tensor& tensor) {
+  return tensor.is_conj();
```
lol

albanD commented on c10/core/DispatchKey.cpp:
```diff
@@ -149,8 +152,6 @@ const char* toString(DispatchKey t) {
     // https://github.com/zou3519/functorch
     // We plan on eventually upstreaming the prototype into core, at which
     // point it will have a different design that should use fewer keys.
-    case DispatchKey::FuncTorchPython:
```
The comment above should go?

anjali411 replied:
yup

anjali411 replied:
actually not sure if the comment above is still valid for the keys below. cc. @zou3519

zou3519 replied:
Please keep the comment

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -1056,6 +1056,24 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  /**
+   Set whether or not the tensor is a zero tensor
+  */
+  void _set_zero(bool value) {
```
In which context do you use this function?

anjali411 replied:
to not propagate the ZT bit in `_like` factory functions

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1,4 +1,5 @@
 import torch
+from torch._C import InferredType
```
can be removed

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +872,59 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+                    val2, res2 = fwAD.unpack_dual(d_o2)
+
+                    if not (res1 is None or res2 is None):
+                        assert torch.equal(res1, res2)
```
You should raise `GradcheckError` here.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1205,6 +1262,13 @@ def _fast_gradcheck(func, func_out, inputs, outputs, eps, rtol,
     inp_tensors_idx, inp_tensors = _get_inp_tensors(inputs)
+    # Backward mode computes v^T * J (VJP)
+    # Since we computed J * u (JVP) through finite difference method, we perform an equality check
+    # between VJP * u, JVP * v
```
nit: "v * JVP"

albanD commented on aten/src/ATen/ZeroTensorFallback.cpp:
```diff
@@ -0,0 +1,99 @@
+
+namespace at {
+
+  // TODO: add a note explaining the design decisions
```
Sounds good

albanD commented on aten/src/ATen/ZeroTensorFallback.cpp:
```diff
@@ -0,0 +1,99 @@
+      }
+    }
+
+    if (is_write.has_value() && !*is_write) {
```
nit: could you add a comment here that this handles view ops?

albanD reviewed (commented):
Looks mostly good to me!

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"




[ghstack-poisoned]

anjali411 commented:
@albanD addressed all the comments, I think!

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

zou3519 commented:
This is a big enough change that I think there is a risk that it breaks functorch and we need to spend a while to figure out how to unbreak it. Could we please confirm that functorch works with this before landing it?

anjali411 commented:
@zou3519 yeah you can check out https://fb.quip.com/xy27ApHtL9Cx

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +613,31 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
+  if (!(self.sizes().size() == 0 || other.sizes().size() == 0)) {
+    TORCH_CHECK(self.sizes() == other.sizes());
+  }
+  TORCH_CHECK(self.device() == other.device());
```
As mentioned offline, this is not true for zero dim Tensors.
I think it is ok here to be more relaxed than the true op if it is hard to replicate the exact same checks.

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +613,31 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
 
+void size_and_device_check(const Tensor& self, const Tensor& other) {
+  // if one of the self or other is a Scalar Tensor
+  if (!(self.sizes().size() == 0 || other.sizes().size() == 0)) {
```
nit: `self.dim() == 0`

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +613,31 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
+  size_and_device_check(self, other);
+  auto commonDtype = at::result_type(self, other);
+  if (self.is_zerotensor()) {
+    return (commonDtype == other.scalar_type()) ? other.mul(alpha) : other.to(commonDtype).mul(alpha);
```
For the second case where the to() is already returning new memory, you can do `alpha == 1? res: res.mul(alpha)` to save an extra copy in the common case where alpha=1.

albanD commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +613,31 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
+  if (self.is_zerotensor()) {
+    return (commonDtype == other.scalar_type()) ? other.mul(alpha) : other.to(commonDtype).mul(alpha);
+  } else {
+    return (commonDtype == self.scalar_type()) ? self.clone() : self.to(commonDtype);
```
Would it be safer to use `_to_copy()` here instead of `to`? Or are we guaranteed that if the dtype is not the same, to will always copy?

albanD commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -352,6 +352,7 @@ Tensor empty_like(
   // never propagate Conjugate and Negative dispatch key
   result._set_conj(false);
   result._set_neg(false);
+  result._set_zero(false);
```
update comment above.

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -767,6 +767,9 @@ Tensor make_qtensor(const Tensor& self, IntArrayRef size, IntArrayRef stride, Qu
 
 Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef stride, optional<int64_t> storage_offset_) {
+  if (self.is_zerotensor()) {
+    return at::_efficientzerotensor(self.sizes(), self.options());
```
Should this return a Tensor of the size `size` and not self's ?

Do we have tests for these?

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1024,6 +1027,11 @@ Tensor alias_with_sizes_and_strides(
     const Vec& strides) {
+  // it's okay to return a new tensor here since we disallow in-place operation on ZeroTensors
+  if (self.is_zerotensor()) {
+    return at::_efficientzerotensor(self.sizes(), self.options());
```
sizes?

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4649,6 +4656,8 @@
   device_check: NoCheck
   device_guard: False
 
+- func: _efficientzerotensor(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
I think you want to do the same as other factory functions:
- make it CompositeExplicitAutograd
- add it at the bottom of derivatives.yaml and make the output non-differentiable.

anjali411 replied:
Since it doesn't have a tensor input, I think I am not allowed to make it CompositeExplicitAutograd. Since now we also support forward mode AD, it might be useful to allow this to specify the output differentiability

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +871,61 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+            dual_inputs.append(inp)
+
+        for i, (fw_grad, u) in enumerate(zip(fw_grads, all_u)):
+            fw_grad.copy_(u.view_as(fw_grad))
```
This can be done in the loop above instead of giving `zeros_like()` to make_dual no?

anjali411 replied:
I don't think so because inp and fw_grad size are not the same ...

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +871,61 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+            fw_grad.copy_(u.view_as(fw_grad))
+
+        dual_inputs_idx = 0
+        for inp in inputs:
```
use tensor_inputs here so that you can remove the if statement below and use `enumerate` to get dual_inputs_idx ?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +871,61 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+                    val1, res1 = fwAD.unpack_dual(d_o1)
+                    val2, res2 = fwAD.unpack_dual(d_o2)
+
+                    if not (res1 is None or res2 is None):
```
nit check that val1/val2 match as well?

anjali411 replied:
that check doesn't need to be in this test

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +871,61 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+
+                    if not (res1 is None or res2 is None):
+                        if not torch.equal(res1, res2):
+                            raise GradcheckError("Mismatch in tangent values for output with index: ", index_o,
```
nit: print the values of res1 res2 here to make debugging easier?

albanD reviewed (commented):
Mostly minor comments.

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

albanD reviewed (approved):
Not 100% about the meta device. But sounds good for now!

bdhirsh commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +614,64 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
+  auto device_ = Device(DeviceType::Meta);
+  auto self_device = self.device();
+  auto meta_out = at::meta::mul(self.to(device_), other.to(device_));
+  if (self.is_meta() || other.is_meta()) {
```
Is "zero tensors on the meta backend" a real use case?

I think there's probably not any reason to add special handling for the meta backend here, since `_efficientzerotensor` looks like it will do all of the right things anyway (and not be very expensive, since it also doesn't allocate data).

Also, totally optional but - if you want to eek out a bit of extra performance, you should be able to do something like:
```
auto meta_out = at::meta::mul(at::native::empty_strided_meta(self.sizes(), self.strides(), self.options()), at::native::empty_strided_meta(other.sizes(), other.strides(), other.options()));
```
to avoid redispatches in the `.to()` call.
```

anjali411 replied:
makes sense! removed it

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -613,6 +614,42 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
+  };
+
+  if (self._is_zerotensor()) {
+    auto res = get_out_like(other);
```
Natalia - what if both self and other are ZTs?

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -318,7 +318,8 @@
 FW_DERIVATIVE_DEFINED_GRAD_TEMPLATE = CodeTemplate("""\
 auto ${inp}_t_raw = toNonOptFwGrad(${inp});
-auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::zeros_like(toNonOptTensor(${inp}));
+auto ${inp}_tensor = toNonOptTensor(${inp});
+auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::_efficientzerotensor(${inp}_tensor.sizes(), ${inp}_tensor.options());
```
Does this PR only use efficient zero tensors in forward-mode AD? Are you planning on using them for reverse-mode AD as well?

anjali411 replied:
> Does this PR only use efficient zero tensors in forward-mode AD? 

yes

> Are you planning on using them for reverse-mode AD as well? 

There aren't that many cases where this would be useful in reverse-mode AD since very few functions have multiple outputs. But, yeah possibly! I see no reason not to.

> Where else are we planning on using efficient zeros tensors in the future?

Alban and I have discussed exposing this in public API via kwarg arg in `torch.zeros` but we haven't decided anything yet.

zou3519 replied:
After reading through this PR: we're going to need to modify functorch to handle efficient zero tensors correctly. In particular, whenever we create a wrapper Tensor, it should receive the zeros dispatch key, otherwise it will not report `_is_zerotensor` correctly.

There are implications for this for tensor subclassing too. In addition to sizes/strides being metadata on a wrapper tensor subclass, the wrapper tensor subclass also probably wants to be able to say that it has the {conjugate, neg, efficient zeros} bit set. (cc @albanD does that sound right?)

If we agree on the above two items then I think that's all we need to say that "efficient zeros tensor works with functorch" and I can fix the problem mentioned in the first paragraph when this PR goes in.

anjali411 replied:
> There are implications for this for tensor subclassing too.

Conjugation and ZT info are not tensor metadata, so I would expect it to work fine for tensor subclasses, if we propagate all the dispatch keys correctly when making tensor subclasses. So, then, this seems like a big limitation of `_make_wrapper_subclass` since it makes tensor subclasses non-composable with any dispatch key feature. 

I was wondering what if we instead add `_make_wrapper_subclass_like` which will have `_like` API semantics, and will allow us to propagate the original tensor's dispatch keys and set them on the wrapper tensor as well, with the option to override the tensor metadata through keyword arguments.

zou3519 replied:
If we go the `_make_wrapper_subclass_like` we should probably stop recommending people use `_make_wrapper_subclass` because that's a good way to shoot oneself in the foot

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -316,9 +316,11 @@
 auto ${inp}_t_raw = toNonOptFwGrad(${inp});
-auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::zeros_like(toNonOptTensor(${inp}));
+auto ${inp}_tensor = toNonOptTensor(${inp});
+auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::zeros({}, ${inp}_tensor.options()).expand(${inp}_tensor.sizes());
```
cc. @albanD this works

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -316,10 +316,11 @@
 auto ${inp}_tensor = toNonOptTensor(${inp});
-auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::_efficientzerotensor(${inp}_tensor.sizes(), ${inp}_tensor.options());
+auto ${inp}_t = ${inp}_t_raw.defined() ? ${inp}_t_raw : at::zeros({}, ${inp}_tensor.options()).expand(${inp}_tensor.sizes());
 """)
```
This needs to be changed back so we can merge it right?

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32144240](https://our.internmc.facebook.com/intern/diff/D32144240)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32144240).

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32834987).

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

albanD commented:
What is the reason for the revert and the update?

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

anjali411 committed:
Update on "Add efficient zero tensors"


Differential Revision: [D32834987](https://our.internmc.facebook.com/intern/diff/D32834987)

[ghstack-poisoned]

vfdev-5 commented on torch/autograd/gradcheck.py:
```diff
@@ -871,8 +871,59 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
+                val2, res2 = fwAD.unpack_dual(d_o2)
+
+                if not (res1 is None or res2 is None):
+                    if not torch.equal(res1, res2):
```
@anjali411 shouldn't we use allclose or _allclose_with_type_promotion here instead of `torch.equal` ?

anjali411 replied:
yeah I think that makes sense!would you like to create a PR with that change? feel free to add me as a reviewer

vfdev-5 replied:
Sure, i'll send a quick patch and add you as a reviewer
